### PR TITLE
feat(nnx): add out_sharding to recurrent cells

### PR DESCRIPTION
**Pull request by aarushisingh04**
## What does this PR do?

adds `out_sharding` parameter to the `__call__` methods of recurrent cell classes in `flax/nnx/nn/recurrent.py`: `LSTMCell`, `OptimizedLSTMCell`, `SimpleCell`, and `GRUCell`.

split from [#5179](https://github.com/google/flax/pull/5179) for easier review.

**changes:**
- adds `*, out_sharding=None` to `__call__` of all 4 cell types
- forwards `out_sharding` to all internal `Linear.__call__()` invocations
- adds `__call__` docstrings with `out_sharding` documented for `SimpleCell` 

**tests:**
- adds `test_out_sharding_signature` for `SimpleCell` in `recurrent_test.py`

## Checklist
- [ ] This PR fixes a minor issue (e.g.: typo or small bug) or improves the docs (you can dismiss the other checks if that's the case).
- [ ] This change is discussed in a Github issue/[discussion](https://github.com/google/flax/discussions) (please add a link).
- [x] The documentation and docstrings adhere to the [documentation guidelines](https://github.com/google/flax/blob/main/docs/README.md#how-to-write-code-documentation).
- [x] This change includes necessary high-coverage tests. (No quality testing = no merge!)

### CODE REVIEW

**Review by samanklesaria — APPROVED** [review]

**Review by IvyZX — APPROVED** [review]

File: tests/nnx/nn/recurrent_test.py [hunk 1]
```diff
@@ -642,5 +642,26 @@ def __call__(self, x):
...
+    ]
+
+    for cell_cls, kwargs in cell_types:
+      with self.subTest(cell_cls=cell_cls.__name__):
+        model = cell_cls(**kwargs, rngs=rngs)
+        carry = model.initialize_carry((1, 4), rngs=rngs)
+        x = jnp.ones((1, 4))
+        # Just verify it accepts out_sharding=None without error
+        out = model(carry, x, out_sharding=None)
+        self.assertLen(out, 2)  # All cells return (new_carry, output) or equivalent structure
```
Comment (IvyZX) [inline tests/nnx/nn/recurrent_test.py]: This test is not valid because it doesn't test when out_sharding has a value, which is the point of this PR.

You can use https://docs.jax.dev/en/latest/jax.sharding.html#jax.sharding.Sharding.is_equivalent_to to check sharding.

**Review by IvyZX — CHANGES_REQUESTED** [review]

File: tests/nnx/nn/recurrent_test.py (hunk 1 above)
Comment (aarushisingh04) [reply tests/nnx/nn/recurrent_test.py]: thanks for catching that, i have made the requested change

File: tests/nnx/nn/recurrent_test.py [hunk 2]
```diff
@@ -642,5 +642,36 @@ def __call__(self, x):
     self.assertEqual(model.lstm.cell.recurrent_dropout.rngs.count[...], 1)
 
 
+class TestCellSharding(absltest.TestCase):
+  def test_out_sharding(self):
+    """Test that out_sharding correctly applies sharding to cell outputs."""
+    rngs = nnx.Rngs(0)
+    devices = np.array(jax.devices())
+    mesh = jax.sharding.Mesh(devices, axis_names=('x',))
+    sharding = jax.sharding.NamedSharding(mesh, jax.sharding.PartitionSpec())
```
Comment (IvyZX) [inline tests/nnx/nn/recurrent_test.py]: `jax.sharding.PartitionSpec()` is the default sharding, aka. everything is replicated & nothing sharded. Using this as test doesn't make sense.

**Review by IvyZX — CHANGES_REQUESTED** [review]

File: tests/nnx/nn/recurrent_test.py (hunk 2 above)
Comment (aarushisingh04) [reply tests/nnx/nn/recurrent_test.py]: applied the fix here, changed `PartitionSpec()` → `PartitionSpec('x',)` to actually shard along the mesh axis, set `batch_size = jax.device_count()` so the axis can be split. 

also added `AxisType.Explicit` + `with mesh:` for correct axis resolution.
